### ISSUE TITLE
Render Tooltip to React Portal to Prevent Clipping

### ISSUE THREAD

**Issue by Eric24** [recharts/recharts#2458]
- [x] I have searched the [issues](https://github.com/recharts/recharts/issues) of this repository and believe that this is not a duplicate.

### What problem does this feature solve?
Currently, if there are enough data series shown in the Tooltip to result in a Tooltip taller than the height of the chart and it's "chrome" (axes, legend, etc.), the tooltip is often clipped by the DOM element that contains the chart (note: it is not the chart itself that is clipping the Tooltip, but rarely if ever will a chart be the only element on a page by itself). By using a React Portal (https://reactjs.org/docs/portals.html), the Tooltip can be rendered independently of the chart and other elements in the DOM hierarchy, thus preventing the clipping.

### What does the proposed API look like?
No change to the API would be necessary. This should be a transparent, non-breaking change.

**Comment by mrcljx** [comment]
For people who are looking for a baseline implementation (with `react-popper`):

```tsx
import { ReactNode, useEffect, useMemo, useRef, useState } from "react";
import { createPortal } from "react-dom";
import { usePopper } from "react-popper";
import { useEvent } from "react-use";
import type { TooltipProps } from "recharts/types/component/Tooltip";

function MyComponent() {
  return (
    <BarChart>
      <Tooltip content={<TooltipContent />} />
    </BarChart>
  }
}

function TooltipContent({ active }: TooltipProps) {
  return (
    <PopperPortal active={active}>
      CUSTOM TOOLTIP RENDERER
    </PopperPortal>
  );
}

function PopperPortal({
  active = true,
  children,
}: {
  active?: boolean;
  children: ReactNode;
}) {
  const [portalElement, setPortalElement] = useState<HTMLDivElement>();
  const [popperElement, setPopperElement] = useState<HTMLDivElement | null>();
  const virtualElementRef = useRef({
    getBoundingClientRect: INITIAL_GET_BOUNDING_CLIENT_RECT,
  });

  const { styles, attributes, update } = usePopper(
    virtualElementRef.current,
    popperElement,
    POPPER_OPTIONS
  );

  useEffect(() => {
    const el = document.createElement("div");
    document.body.appendChild(el);
    setPortalElement(el);
    return () => el.remove();
  }, []);

  useEvent("mousemove", ({ clientX: x, clientY: y }) => {
    if (!active) return;
    virtualElementRef.current.getBoundingClientRect =
      generateGetBoundingClientRect(x, y);
    update?.();
  });

  if (!portalElement) return null;

  return createPortal(
    <div
      ref={setPopperElement}
      {...attributes.popper}
      style={{
        ...styles.popper,
        zIndex: 1000,
        display: active ? "block" : "none",
      }}
    >
      {children}
    </div>,
    portalElement
  );
}
```

**Comment by reify-thomas-smith** [comment]
As an alternative, `<Tooltip />` could add a `portal` prop taking a DOM element, and render its contents inside that element if it is given.

**Comment by Sin9k** [comment]
We need Portals! Please support it!

**Comment by CyrilQuandalle** [comment]
Hello,
Has this feature request development been shipped ?
Thanks in advance for your answer.

**Comment by ckifer** [comment]
No, no work on portals yet. Refactoring needs to be done first in order to make it viable

**Comment by CyrilQuandalle** [comment]
ok thanks for the answer, then external custom Tooltip it is ;-)

**Comment by dlvhdr** [comment]
@CyrilQuandalle could you share how you've solved it? (your custom implementation)

I'm facing this issue as well as another issue where the tooltip hides the entire chart. It happens if it's big enough, the chart is small enough and is located near the edge of the viewport. I would expect it to go above the chart in that case and not hide it. 

All in all, seems like the tooltip's implementation is very lacking.

**Comment by PavelVanecek** [comment]
This should be available in 3.x

**Comment by PavelVanecek** [comment]
I am going to close this as we're getting closer to the 3.0 release.

**Comment by PavelVanecek** [comment]
Hello everyone who's watching this issue! We have released an alpha version `3.0.0-alpha.2` which fixes this problem.

Here is a migration guide: https://github.com/recharts/recharts/wiki/3.0-migration-guide

As of now we have fixed all the new 3.x bugs we are aware of so, and are ready for your feedback. Please feel free to try `recharts: 3.0.0-alpha.2` in a test build and let us know how it went!

**Comment by andhikaribrahim** [comment]
If anyone experiencing this in recharts v2, maybe you can use this as reference (in my case i'm using MUI Portal component or you can create your own):

```typescript
import Portal from '@mui/material/Portal'

const { ..., coordinate } = props

const chartContainer = document.querySelector('.your-chart-container .recharts-wrapper')
const wrapperRect = chartContainer?.getBoundingClientRect()

const pageX = (wrapperRect?.left ?? 0) + (coordinate?.x ?? 0)
const pageY = (wrapperRect?.top ?? 0) + (coordinate?.y ?? 0)

/* rest of your component logics */
return (
  <Portal container={document.body}>
    <Box
      sx={{
        pointerEvents: 'none',
        position: 'fixed',
        top: pageY,
        left: pageX,
        backgroundColor: '#212121',
        zIndex: 102,
      }}
    >
      /* rest of your component */
    </Box>
  </Portal>
)
```

And your tooltip should look like this
```typescript
<Tooltip
  allowEscapeViewBox={{ x: true, y: true }}
  content={<CustomTooltip />}
  cursor={false}
  wrapperStyle={{
    pointerEvents: 'none',
  }}
/>
```
